man/auracle.1.pod
=head1 NAME

auracle - an AUR client with a focus on flexibility

=head1 SYNOPSIS

I<auracle> [OPTIONS...] COMMAND [ARGS...]

=head1 DESCRIPTION

auracle is a tool for querying information stored by the Arch User Repository
(AUR). Invoking auracle consists of supplying an operation, any applicable
options, and usually one or more arguments.

=head1 OPTIONS

=over 4

=item B<--color=>

Control colored output. Argument must be one of I<auto>, I<never>, or
I<always>. When set to I<auto>, color will be enabled if stdout is connected
to a terminal.

This option defaults to I<auto>.

=item B<-h>, B<--help>

Print a short help text and exit.

=item B<--literal>

When used with the B<search> command, interpret all search terms as literal,
rather than as regular expressions.

=item B<--quiet>

When used with the B<search> and B<outdated> commands, output will be limited to
package names only.

=item B<-r>, B<--recurse>

When used with the B<clone> command, recursively follow and clone
dependencies of each given argument.

=item B<--resolve-deps=>I<DEPLIST>

When performing recursive operations, control the kinds of dependencies that
auracle will consider. By default, this is all dependency kinds -- I<depends>,
I<makedepends>, and I<checkdepends>. This flag takes a comma-delimited string
value with one to many of these dependency kinds.

Additionally, some special syntax exists to make this more composable. A
leading I<!> or I<^> character will cause dependencies to be removed from the
set, whereas a leading I<+> will cause dependencies to be added to the set.
These special indicators are only understood when they appear as the first
character in the string and never after a comma.

Specifying this flag multiple times is additive, for example:

  auracle buildorder --resolve-deps=depends --resolve-deps=+makedepends

Is the same as I<depends,makedepends> or I<^checkdepends>.

=item B<--searchby=>I<BY>

When used with the B<search> and B<rawsearch> commands, change the search
dimension. I<BY> must be one of I<name>, I<name-desc>, I<maintainer>,
I<depends>, I<makedepends>, I<optdepends>, I<checkdepends>, I<submitter>,
I<provides>, I<conflicts>, I<replaces>, I<keywords>, I<groups>, or
I<comaintainers>.

This option defaults to I<name-desc>.

=item B<--sort=>I<KEY>, B<--rsort=>I<KEY>

For search and info queries, sorts the results in ascending and descending
order, respectively. I<KEY> must be one of: B<name>, B<popularity>, B<votes>,
B<firstsubmitted>, or B<lastmodified>.

This option defaults to sorting by B<name> in ascending order.

=item B<--show-file=>I<FILE>

Name of the file to fetch with the B<show> command.

This option defaults to B<PKGBUILD>.

=item B<-C >I<DIR>, B<--chdir=>I<DIR>

Change directory to I<DIR> before performing any actions. Only useful with the
B<clone> command.

=item B<--proxy>I<URL>

Specifies the URL to a proxy server that can handle the /rpc/v5/info and
/rpc/v5/search endpoints of the AUR. This could be a local implementation
which performs caching or serves from the packages json file.

=back

=head1 COMMANDS

=over 4

=item B<buildorder> I<PACKAGES>...

Pass one to many arguments to print a build order for the given packages.  The
resulting output will be the total ordering to build all packages. Each line is
at least two columns delimited by a single whitespace. The first column contains an
identifier and the second column names the dependency. Possible identifiers
are:

=over 4

B<AUR>        The dependency was found in the AUR and needs to be built and
installed. The output will contain an additional column of the pkgbase the
package is provided by.

B<REPOS>      The dependency was found in a binary repo and can be installed via pacman.

B<UNKNOWN>    The dependency was unable to be found anywhere. This might
indicate a broken dependency chain. The output will contain additional columns
denoting the dependency path that was walked (in reverse) to get to the missing
dependency.

=back

Additionally, both of B<AUR> and B<REPOS> can be prefixed with B<SATISFIED> to
indicate that the dependency is already installed. B<AUR> may be prefixed with
B<TARGET> to indicate that the package was explicitly specified on the
commandline.

=item B<clone> I<PACKAGES>...

Pass one to many arguments to get clone git repositories. Use the
B<--recurse> flag to get dependencies of packages. If the git repository
already exists, the repository will instead be updated. A best effort attempt
will be made to preserve existing local changes to packages. Auracle will
instruct git to autostash staged changes and will then rebase new incoming
changes before unstashing. This is intended to handle the case of makepkg
updating the I<pkgver> attribute of PKGBUILDs but may not handle more complex
changes made by a local user.

=item B<info> I<PACKAGES>...

Pass one to many arguments to perform an info query.

=item B<rawinfo> I<PACKAGES>...

Dump the raw JSON response from the AUR for an info request.

=item B<rawsearch> I<TERMS>...

Dump the raw JSON response from the AUR for search requests formed from the
given terms.

=item B<search> I<TERMS>...

Pass one to many arguments to perform a search query. Results will be the
intersection of all terms. Terms may be a regular expression as described by
the POSIX extended regex specification.

B<NOTE>: the AUR does not actually support searching by regular expressions
and support in auracle is implemented on a best-effort basis.

=item B<resolve> I<TERMS>...

Pass one to many dependency string arguments to perform a search query for
packages that can satisfy the input dependencies. Dependency syntax follows
that of pacman and makepkg, for example: I<auracle> or I<auracleE<gt>=21>.

=item B<show> I<PACKAGES>...

Pass one to many arguments to print source files for the given packages. The
file fetched is controlled by the B<--show-file> flag.

=item B<outdated> [I<PACKAGES>...]

Pass one to many arguments to check for newer versions existing in the AUR.
Each argument is assumed to be a package installed locally using B<pacman>(8).
If no arguments are given, pacman is queried for all foreign packages as an
input to this operation.

=item B<update> [I<PACKAGES>...]

Similar to B<outdated>, but download packages found to be outdated. Pass the
B<--recurse> flag to download new dependencies of packages as well.

=back

=head1 CUSTOM FORMATTING

Auracle provides the ability to customize the output format from B<info> and
B<search> operations. The format specification is similar to that of python.
Each field is wrapped in curly brackets and is optionally suffixed with a colon
and additional formatting directives.

B<NOTE:> not all fields are available for B<search> queries.

String fields: B<{name}>, B<{version}>, B<{description}>, B<{submitter}>,
B<{maintainer}>, B<{pkgbase}>, B<{url}>.

Floating point fields: B<{votes}>, B<{popularity}>.

Time fields: B<{submitted}>, B<{modified}>, B<{outofdate}>. These fields
default to using your locale's default date and time, but can be further
customized in their format, according to the specifier described in
strftime(3), For example, B<{submitted:%Y.%m.%d}> would show year month and
day in dotted form.

String array fields: B<{depends}>, B<{makedepends}>, B<{checkdepends}>,
B<{optdepends}>, B<{conflicts}>, B<{provides}>, B<{replaces}>, B<{groups}>,
B<{keywords}>, B<{licenses}>, B<{comaintainers}>. These fields default to using
a delimiter of two whitespace, but can be customized, e.g.  B<{depends:,}> to
comma delimit the values of the depends field.

Example: recreating auracle's search output:

  auracle search -F $'aur/{name} ({votes}, {popularity})\n    {description}' ...

Example: listing dependencies of packages:

  auracle info -F '{depends} {makedepends} {checkdepends}' ...

=head1 AUTHOR

Dave Reisner E<lt>d@falconindy.comE<gt>



src/auracle/auracle.cc
// SPDX-License-Identifier: MIT
#include "auracle.hh"

#include <cerrno>
#include <filesystem>
#include <functional>
#include <iostream>
#include <regex>
#include <string_view>
#include <tuple>

#include "absl/algorithm/container.h"
#include "absl/container/flat_hash_set.h"
#include "absl/status/statusor.h"
#include "aur/response.hh"
#include "auracle/dependency.hh"
#include "auracle/format.hh"
#include "auracle/pacman.hh"
#include "auracle/search_fragment.hh"
#include "auracle/sort.hh"

namespace fs = std::filesystem;

using SearchBy = aur::SearchRequest::SearchBy;

namespace auracle {

namespace {

// ResponseMerger allows us to associate multiple RPC requests and issue
// a callback only after they all succeed.
class ResponseMerger {
 public:
  using MergedResultCallback =
      absl::AnyInvocable<void(absl::StatusOr<std::vector<aur::Package>>) &&>;

  static ResponseMerger* New(MergedResultCallback&& callback) {
    return new ResponseMerger(std::move(callback));
  }

  ~ResponseMerger() {
    if (status_.ok()) {
      std::move(callback_)(std::move(packages_));
    } else {
      std::move(callback_)(status_);
    }
  }

  aur::Client::RpcResponseCallback callback() {
    ++inflight_calls_;

    return [this](absl::StatusOr<aur::RpcResponse> response) {
      status_.Update(response.status());
      if (status_.ok()) {
        absl::c_move(response->packages, std::back_inserter(packages_));
      }

      if (--inflight_calls_ == 0) {
        delete this;
      }

      return 0;
    };
  }

 private:
  ResponseMerger(MergedResultCallback callback)
      : callback_(std::move(callback)) {}

  MergedResultCallback callback_;
  int inflight_calls_ = 0;

  absl::Status status_;
  std::vector<aur::Package> packages_;
};

int ErrorNotEnoughArgs() {
  std::cerr << "error: not enough arguments.\n";
  return -EINVAL;
}

void FormatLong(const std::vector<aur::Package>& packages,
                const auracle::Pacman* pacman) {
  for (const auto& p : packages) {
    format::Long(p, pacman->GetLocalPackage(p.name));
  }
}

void FormatNameOnly(const std::vector<aur::Package>& packages) {
  for (const auto& p : packages) {
    format::NameOnly(p);
  }
}

void FormatShort(const std::vector<aur::Package>& packages,
                 const auracle::Pacman* pacman) {
  for (const auto& p : packages) {
    format::Short(p, pacman->GetLocalPackage(p.name));
  }
}

void FormatCustom(const std::vector<aur::Package>& packages,
                  const std::string& format) {
  for (const auto& p : packages) {
    format::Custom(format, p);
  }
}

void SortUnique(std::vector<aur::Package>& packages,
                const sort::Sorter& sorter) {
  absl::c_sort(packages, sorter);
  packages.resize(std::unique(packages.begin(), packages.end()) -
                  packages.begin());
}

std::vector<std::string> NotFoundPackages(
    const std::vector<std::string>& want, const std::vector<aur::Package>& got,
    const auracle::PackageCache& package_cache) {
  std::vector<std::string> missing;

  for (const auto& name : want) {
    if (package_cache.LookupByPkgname(name) != nullptr) {
      continue;
    }

    if (absl::c_find_if(got, [&](const aur::Package& pkg) {
          return pkg.name == name;
        }) != got.end()) {
      continue;
    }

    missing.push_back(name);
  }

  return missing;
}

bool ChdirIfNeeded(const fs::path& target) {
  if (target.empty()) {
    return true;
  }

  std::error_code ec;
  fs::current_path(target, ec);
  if (ec.value() != 0) {
    std::cerr << "error: failed to change directory to " << target << ": "
              << ec.message() << "\n";
    return false;
  }

  return true;
}

bool RpcResponseIsFailure(const absl::StatusOr<aur::RpcResponse>& response) {
  if (response.ok()) {
    return false;
  }

  std::cerr << "error: " << response.status() << '\n';
  return true;
}

}  // namespace

Auracle::Auracle(Options options)
    : client_(aur::Client::New(aur::Client::Options()
                                   .set_baseurl(options.baseurl)
                                   .set_proxy(options.proxy)
                                   .set_useragent("Auracle/" PROJECT_VERSION))),
      pacman_(options.pacman) {}

void Auracle::ResolveMany(const std::vector<std::string>& depstrings,
                          aur::Client::RpcResponseCallback callback) {
  // A naive implementation of ResolveMany could be just calling search+info in
  // a loop, but we make this more complicated such that for N arguments, we can
  // issue N search requests and a single info request, rather than as many as
  // N*2 requests.

  auto deps = std::make_shared<std::vector<Dependency>>();

  auto* merger = ResponseMerger::New(
      [this, deps, callback = std::move(callback)](
          absl::StatusOr<std::vector<aur::Package>> packages) mutable {
        if (!packages.ok() || packages->empty()) {
          return std::move(callback)(packages);
        }

        aur::InfoRequest info_request;
        for (const auto& pkg : *packages) {
          info_request.AddArg(pkg.name);
        }

        client_->QueueRpcRequest(
            info_request,
            [&, callback = std::move(callback),
             deps](absl::StatusOr<aur::RpcResponse> info_response) mutable {
              if (info_response.ok()) {
                std::erase_if(info_response->packages,
                              [&](const aur::Package& package) {
                                for (const auto& dep : *deps) {
                                  if (dep.SatisfiedBy(package)) {
                                    return false;
                                  }
                                }
                                return true;
                              });
              }

              return std::move(callback)(std::move(info_response));
            });

        return 0;
      });

  for (const auto& depstring : depstrings) {
    client_->QueueRpcRequest(
        aur::SearchRequest(SearchBy::PROVIDES,
                           deps->emplace_back(depstring).name()),
        merger->callback());
  }
}

void Auracle::IteratePackages(std::vector<std::string> args,
                              Auracle::PackageIterator* state) {
  std::erase_if(args, [&](const std::string& arg) {
    return state->package_cache.LookupByPkgname(arg) != nullptr;
  });
  if (args.empty()) {
    return;
  }

  aur::InfoRequest info_request(args);
  client_->QueueRpcRequest(
      info_request, [this, state, want = std::move(args)](
                        absl::StatusOr<aur::RpcResponse> response) {
        if (RpcResponseIsFailure(response)) {
          return -EIO;
        }

        auto& results = response.value().packages;

        for (const auto& p :
             NotFoundPackages(want, results, state->package_cache)) {
          if (!pacman_->HasPackage(p)) {
            std::cerr << "no results found for " << p << "\n";
          }
        }

        for (auto& result : results) {
          // check for the pkgbase existing in our repo
          const bool have_pkgbase =
              state->package_cache.LookupByPkgbase(result.pkgbase) != nullptr;

          // Regardless, try to add the package, as it might be another member
          // of the same pkgbase.
          auto [p, added] = state->package_cache.AddPackage(std::move(result));

          if (!added || have_pkgbase) {
            continue;
          }

          if (state->callback) {
            state->callback(*p);
          }

          if (state->recurse) {
            std::vector<std::string> alldeps;
            alldeps.reserve(p->depends.size() + p->makedepends.size() +
                            p->checkdepends.size());

            for (auto kind : state->resolve_depends) {
              for (const auto& dep : GetDependenciesByKind(p, kind)) {
                alldeps.push_back(Dependency(dep).name());
              }
            }

            if (alldeps.empty()) {
              return 0;
            }

            IteratePackages(std::move(alldeps), state);
          }
        }

        return 0;
      });
}

int Auracle::Info(const std::vector<std::string>& args,
                  const CommandOptions& options) {
  if (args.empty()) {
    return ErrorNotEnoughArgs();
  }

  std::vector<aur::Package> packages;
  client_->QueueRpcRequest(aur::InfoRequest(args),
                           [&](absl::StatusOr<aur::RpcResponse> response) {
                             if (RpcResponseIsFailure(response)) {
                               return -EIO;
                             }

                             using std::swap;
                             swap(packages, response.value().packages);

                             return 0;
                           });

  auto r = client_->Wait();
  if (r < 0) {
    return r;
  }

  if (packages.empty()) {
    return -ENOENT;
  }

  // It's unlikely, but still possible that the results may not be unique when
  // our query is large enough that it needs to be split into multiple requests.
  SortUnique(packages, options.sorter);

  if (!options.format.empty()) {
    FormatCustom(packages, options.format);
  } else {
    FormatLong(packages, pacman_);
  }

  return 0;
}

int Auracle::Resolve(const std::vector<std::string>& args,
                     const CommandOptions& options) {
  if (args.empty()) {
    return ErrorNotEnoughArgs();
  }

  std::vector<aur::Package> providers;

  ResolveMany(args, [&](absl::StatusOr<aur::RpcResponse> response) {
    if (RpcResponseIsFailure(response)) {
      return -EIO;
    }

    absl::c_move(response->packages, std::back_inserter(providers));
    return 0;
  });

  int r = client_->Wait();
  if (r < 0) {
    return r;
  }

  SortUnique(providers, options.sorter);

  if (!options.format.empty()) {
    FormatCustom(providers, options.format);
  } else if (options.quiet) {
    FormatNameOnly(providers);
  } else {
    FormatShort(providers, pacman_);
  }

  return 0;
}

int Auracle::Search(const std::vector<std::string>& args,
                    const CommandOptions& options) {
  if (args.empty()) {
    return ErrorNotEnoughArgs();
  }

  std::vector<std::regex> patterns;
  for (const auto& arg : args) {
    try {
      patterns.emplace_back(arg, std::regex::icase | std::regex::optimize);
    } catch (const std::regex_error& e) {
      std::cerr << "error: invalid regex: " << arg << "\n";
      return -EINVAL;
    }
  }

  const auto matches = [&](const aur::Package& p) {
    return absl::c_all_of(patterns, [&](const std::regex& re) {
      switch (options.search_by) {
        case SearchBy::NAME:
          return std::regex_search(p.name, re);
        case SearchBy::NAME_DESC:
          return std::regex_search(p.name, re) ||
                 std::regex_search(p.description, re);
        default:
          // The AUR only matches maintainer and *depends
          // fields exactly so there's no point in doing
          // additional filtering on these types.
          return true;
      }
    });
  };

  // 'name' and 'name-desc' are the only dimensions where the AUR allows
  // substring matching, so that's the only case where we're able to provide
  // something resembling regex support.
  const bool allow_regex =
      options.allow_regex && (options.search_by == SearchBy::NAME ||
                              options.search_by == SearchBy::NAME_DESC);

  std::vector<aur::Package> packages;
  for (const auto& arg : args) {
    std::string_view frag = arg;
    if (allow_regex) {
      frag = GetSearchFragment(arg);
      if (frag.empty()) {
        std::cerr << "error: search string '" << arg
                  << "' insufficient for searching by regular expression.\n";
        return -EINVAL;
      }
    }

    client_->QueueRpcRequest(aur::SearchRequest(options.search_by, frag),
                             [&](absl::StatusOr<aur::RpcResponse> response) {
                               if (RpcResponseIsFailure(response)) {
                                 return -EIO;
                               }

                               auto& results = response.value().packages;
                               std::copy_if(
                                   std::make_move_iterator(results.begin()),
                                   std::make_move_iterator(results.end()),
                                   std::back_inserter(packages), matches);
                               return 0;
                             });
  }

  int r = client_->Wait();
  if (r < 0) {
    return r;
  }

  SortUnique(packages, options.sorter);

  if (!options.format.empty()) {
    FormatCustom(packages, options.format);
  } else if (options.quiet) {
    FormatNameOnly(packages);
  } else {
    FormatShort(packages, pacman_);
  }

  return 0;
}

int Auracle::Clone(const std::vector<std::string>& args,
                   const CommandOptions& options) {
  if (args.empty()) {
    return ErrorNotEnoughArgs();
  }

  if (!ChdirIfNeeded(options.directory)) {
    return -EINVAL;
  }

  int ret = 0;
  PackageIterator iter(
      options.recurse, options.resolve_depends,
      [this, &ret](const aur::Package& p) {
        client_->QueueCloneRequest(
            aur::CloneRequest(p.pkgbase),
            [&ret,
             pkgbase = p.pkgbase](absl::StatusOr<aur::CloneResponse> response) {
              if (response.ok()) {
                std::cout << response.value().operation << " complete: "
                          << (fs::current_path() / pkgbase).string() << "\n";
              } else {
                std::cerr << "error: clone failed for " << pkgbase << ": "
                          << response.status() << "\n";
                ret = -EIO;
              }
              return 0;
            });
      });

  IteratePackages(args, &iter);

  int r = client_->Wait();
  if (r < 0) {
    return r;
  }

  if (iter.package_cache.empty()) {
    return -ENOENT;
  }

  return ret;
}

int Auracle::Show(const std::vector<std::string>& args,
                  const CommandOptions& options) {
  if (args.empty()) {
    return ErrorNotEnoughArgs();
  }

  int resultcount = 0;
  client_->QueueRpcRequest(
      aur::InfoRequest(args), [&](absl::StatusOr<aur::RpcResponse> response) {
        if (RpcResponseIsFailure(response)) {
          return -EIO;
        }

        resultcount = response.value().packages.size();

        for (const auto& pkg : response.value().packages) {
          client_->QueueRawRequest(
              aur::RawRequest::ForSourceFile(pkg, options.show_file),
              [&options, print_header = resultcount > 1, pkgbase = pkg.pkgbase](
                  absl::StatusOr<aur::RawResponse> response) {
                if (absl::IsNotFound(response.status())) {
                  std::cerr << "error: file '" << options.show_file
                            << "' not found for package '" << pkgbase << "'\n";
                  return -ENOENT;
                } else if (!response.ok()) {
                  std::cerr << "error: request failed: " << response.status()
                            << "\n";
                  return -EIO;
                }

                if (print_header) {
                  std::cout << "### BEGIN " << pkgbase << "/"
                            << options.show_file << "\n";
                }
                std::cout << response.value().bytes << "\n";
                return 0;
              });
        }
        return 0;
      });

  int r = client_->Wait();
  if (r < 0) {
    return r;
  }

  if (resultcount == 0) {
    return -ENOENT;
  }

  return 0;
}

int Auracle::BuildOrder(const std::vector<std::string>& args,
                        const CommandOptions& options) {
  if (args.empty()) {
    return ErrorNotEnoughArgs();
  }

  PackageIterator iter(/* recurse = */ true, options.resolve_depends, nullptr);
  IteratePackages(args, &iter);

  int r = client_->Wait();
  if (r < 0) {
    return r;
  }

  if (iter.package_cache.empty()) {
    return -ENOENT;
  }

  std::vector<
      std::tuple<std::string, const aur::Package*, std::vector<std::string>>>
      total_ordering;
  absl::flat_hash_set<std::string> seen;
  for (const auto& arg : args) {
    iter.package_cache.WalkDependencies(
        arg,
        [&](const Dependency& dep, const aur::Package* package,
            const std::vector<std::string>& dependency_path) {
          if (seen.emplace(dep.name()).second) {
            total_ordering.emplace_back(dep.name(), package, dependency_path);
          }
        },
        options.resolve_depends);
  }

  for (const auto& [name, pkg, dependency_path] : total_ordering) {
    const bool satisfied = pacman_->DependencyIsSatisfied(name);
    const bool from_aur = pkg != nullptr;
    const bool unknown = !from_aur && !pacman_->HasPackage(name);
    const bool is_target = absl::c_find(args, name) != args.end();

    if (unknown) {
      std::cout << "UNKNOWN";
      r = -ENXIO;
    } else {
      if (is_target) {
        std::cout << "TARGET";
      } else if (satisfied) {
        std::cout << "SATISFIED";
      }

      if (from_aur) {
        std::cout << "AUR";
      } else {
        std::cout << "REPOS";
      }
    }

    if (unknown) {
      for (auto iter = dependency_path.crbegin();
           iter != dependency_path.crend(); ++iter) {
        std::cout << " " << *iter;
      }
    } else {
      std::cout << " " << name;
      if (from_aur) {
        std::cout << " " << pkg->pkgbase;
      }
    }

    std::cout << "\n";
  }

  return r;
}

int Auracle::Update(const std::vector<std::string>& args,
                    const CommandOptions& options) {
  if (!ChdirIfNeeded(options.directory)) {
    return -EINVAL;
  }

  std::vector<aur::Package> packages;
  auto r = GetOutdatedPackages(args, &packages);
  if (r < 0) {
    return r;
  }

  if (packages.empty()) {
    return -ENOENT;
  }

  int ret = 0;
  PackageIterator iter(
      options.recurse, options.resolve_depends, [&](const aur::Package& p) {
        client_->QueueCloneRequest(
            aur::CloneRequest(p.pkgbase),
            [&ret,
             pkgbase = p.pkgbase](absl::StatusOr<aur::CloneResponse> response) {
              if (response.ok()) {
                std::cout << response.value().operation << " complete: "
                          << (fs::current_path() / pkgbase).string() << "\n";
              } else {
                std::cerr << "error: clone failed for " << pkgbase << ": "
                          << response.status() << "\n";
                ret = -EIO;
              }
              return 0;
            });
      });

  std::vector<std::string> outdated;
  outdated.reserve(packages.size());
  for (const auto& p : packages) {
    outdated.push_back(p.name);
  }

  IteratePackages(outdated, &iter);

  return client_->Wait();
}

int Auracle::GetOutdatedPackages(const std::vector<std::string>& args,
                                 std::vector<aur::Package>* packages) {
  aur::InfoRequest info_request;

  auto local_pkgs = pacman_->LocalPackages();
  for (const auto& pkg : local_pkgs) {
    if (args.empty() || absl::c_find(args, pkg.pkgname) != args.end()) {
      info_request.AddArg(pkg.pkgname);
    }
  }

  client_->QueueRpcRequest(
      info_request, [&](absl::StatusOr<aur::RpcResponse> response) {
        if (RpcResponseIsFailure(response)) {
          return -EIO;
        }

        auto& results = response.value().packages;
        std::copy_if(std::make_move_iterator(results.begin()),
                     std::make_move_iterator(results.end()),
                     std::back_inserter(*packages), [&](const aur::Package& p) {
                       auto local = pacman_->GetLocalPackage(p.name);

                       return local &&
                              Pacman::Vercmp(p.version, local->pkgver) > 0;
                     });

        return 0;
      });

  return client_->Wait();
}

int Auracle::Outdated(const std::vector<std::string>& args,
                      const CommandOptions& options) {
  std::vector<aur::Package> packages;

  auto r = GetOutdatedPackages(args, &packages);
  if (r < 0) {
    return r;
  }

  if (packages.empty()) {
    return -ENOENT;
  }

  // Not strictly needed, but let's keep output order stable
  std::sort(packages.begin(), packages.end(),
            sort::MakePackageSorter("name", sort::OrderBy::ORDER_ASC));

  for (const auto& r : packages) {
    if (options.quiet) {
      format::NameOnly(r);
    } else {
      auto local = pacman_->GetLocalPackage(r.name);
      format::Update(*local, r);
    }
  }

  return 0;
}

namespace {

int RawRequestDone(absl::StatusOr<aur::RawResponse> response) {
  if (!response.ok()) {
    std::cerr << "error: request failed: " << response.status() << "\n";
    return -EIO;
  }

  std::cout << response.value().bytes << "\n";
  return 0;
}

}  // namespace

int Auracle::RawSearch(const std::vector<std::string>& args,
                       const CommandOptions& options) {
  for (const auto& arg : args) {
    client_->QueueRawRequest(aur::SearchRequest(options.search_by, arg),
                             &RawRequestDone);
  }

  return client_->Wait();
}

int Auracle::RawInfo(const std::vector<std::string>& args,
                     const CommandOptions&) {
  client_->QueueRawRequest(aur::InfoRequest(args), &RawRequestDone);

  return client_->Wait();
}

}  // namespace auracle

/* vim: set et ts=2 sw=2: */
